src/snowflake/snowpark/_internal/analyzer/metadata_utils.py

```diff
@@ -182,3 +188,27 @@ def infer_metadata(
             quoted_identifiers = None
 
     return PlanMetadata(attributes=attributes, quoted_identifiers=quoted_identifiers)
+
+
+def cache_metadata_on_select_statement(
+    source_plan: Optional[LogicalPlan], metadata: PlanMetadata
+) -> None:
+    """
+    Cache metadata on a SelectStatement source plan.
+    """
+    from snowflake.snowpark._internal.analyzer.select_statement import SelectStatement
+
+    if (
+        isinstance(source_plan, SelectStatement)
```

Comment:
if the source plan is None, will this throw an error? maybe just do a check of source_plan is None, then return

Reply:
this is not necessary in python because isinstance(source_plan, SelectStatement) will return False if source_plan is None (because it is NoneType)